### PR TITLE
Add west.configuration.Configuration class

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
This is basically a prep work PR which:

- helps us out with #149 
- lays the groundwork for resolving some longstanding TODOs and ugly/buggy code within the west.manifest subpackage

### CODE REVIEW

**Review by tejlmand — APPROVED** [review]
looks good.

Also done some local testing to ensure no behavioral changes.
All seems good :+1:

File: tests/test_manifest.py [hunk 1]
```diff
@@ -112,8 +112,10 @@ def test_project_init():
     assert p.abspath == os.path.join(TOPDIR, 'p')
     assert p.posixpath == TOPDIR_POSIX + '/p'
 
-def test_manifest_from_data():
+def test_manifest_from_data_without_topdir():
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply tests/test_manifest.py:115]: Thanks.

File: src/west/app/main.py [hunk 2]
```diff
@@ -702,7 +704,7 @@ def set_zephyr_base(args, manifest, topdir):
                 pass
             if projects:
                 zephyr = projects[0]
-                config.update_config('zephyr', 'base', zephyr.path)
+                config.set('zephyr.base', zephyr.path)
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply src/west/app/main.py:709]: > with the new Zephyr LTS 2.7 out, that has Zephyr CMake package, should we remove `zephyr.base` from `.west/config` instead ?

Yes, but not for this PR, I think.

File: src/west/configuration.py [hunk 3]
```diff
@@ -70,6 +127,278 @@ class ConfigFile(Enum):
     GLOBAL = 3
     LOCAL = 4
 
+class Configuration:
```
Comment (marc-hb) [inline src/west/configuration.py:131]: `class Configurations`, plural?

Before reading the pydoc carefully I wrongly guessed that there would be 3 separate instances: local, global and system.

File: src/west/app/main.py [hunk 4]
```diff
@@ -90,9 +91,11 @@ def run(self, argv):
 
         # Read the configuration files. We need this to get
         # manifest.path to parse the manifest, etc.
-        #
-        # TODO: re-work to avoid global state (#149).
-        config.read_config(topdir=self.topdir)
+        self.config = west.configuration.Configuration(topdir=self.topdir)
+
+        # Also set up the global configuration object to match, for
```
Comment (marc-hb) [inline src/west/app/main.py:96]:
```suggestion
        # Also set up the deprecated global configuration object to match, for
```
Is it actually deprecated? Not sure yet.

**Review by marc-hb — COMMENTED** [review]
Cosmetic comments only, sorry not familiar enough for more.

Is there any rationale for this change outside testing? I couldn't tell for sure from the commit messages. In more concrete terms: do you ever expect a real-world use case to ever instantiante two _different_ Configuration objects? I would assume Configuration to be immutable (which I think explain why it could stay a global for so long).

**Comment by mbolivar-nordic** [comment]
@marc-hb thanks for review.

It's not immutable -- it has setters, actually :). Maybe you meant some other word.

And there are two use cases to keep in mind:

1. running a west command in at most one workspace from the command line (at most one Configuration)
2. using west as an API, potentially with multiple different workspaces in the picture (as many Configurations as workspaces -- that is why it is a singular 'Configuration', not 'Configurations' -- each workspace gets one)

For 1., what we have now is enough. For 2., it is not, and west is also meant to be used directly via its APIs.

**Pull request merged**